From patchwork Thu Oct 11 18:46:46 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Peter Shier <pshier@google.com>
X-Patchwork-Id: 10637131
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 5EDBB933
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 11 Oct 2018 18:47:00 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 4E2E22BECE
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 11 Oct 2018 18:47:00 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 3F8912BED4; Thu, 11 Oct 2018 18:47:00 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id E0FF62BECE
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 11 Oct 2018 18:46:59 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1729615AbeJLCPZ (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 11 Oct 2018 22:15:25 -0400
Received: from mail-it1-f201.google.com ([209.85.166.201]:50641 "EHLO
        mail-it1-f201.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727944AbeJLCPY (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 11 Oct 2018 22:15:24 -0400
Received: by mail-it1-f201.google.com with SMTP id h20-v6so10323557ita.0
        for <kvm@vger.kernel.org>; Thu, 11 Oct 2018 11:46:57 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=N9lYw2ci+E9PLuVnKlVCCS6Jk1VlwTR1kIz4R216C3A=;
        b=hxJEFBouESW2OKcHY/MIvG379JQ/QH2kq0RFbSAX5BTZ6TI6JZY2+aEOhuwFDRiyqr
         uGUrLwcxEwY4G/u1yKpND1+/eJhHAI5h8cCWD09Kouz3brezI/9BjxawUH+4PbM2LwGI
         ozJUsK/ydDs3+XI19JQBaMyU19m/oil0kA/VIwMkc7v5UaxuS/+2SufHcPh4bQ4xBkq7
         pHh/+GsG675p53YPWMWrWenj0GTDE1GVpjzzuwp3/s9lQnZc+AG0lsMGK82DlRopTuC8
         yqcwEroFyl9DdMbAsRGcP1139QQj0xogs1ZugwazQ7yOfEW5qg0dQE3oGg9ShYUYzdPN
         BOYA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=N9lYw2ci+E9PLuVnKlVCCS6Jk1VlwTR1kIz4R216C3A=;
        b=NRl+TSFk4ts1xYP/T0nNSHUSBBsqg4PZNCYrtAVV1A/M9cNS8H3Z8N5HX7JRpLve2j
         Wctn8nMYqMPwyrVbVKiaijCXQwytMfrNWn/8KeNiDp2wh7hViprNusypsQdbir52DAF9
         B5Cl8gwaBmgFhGzqWDfCpA17/FAjBL0JPbai3T7idBl5uHWOy2uvFfugglA4lYdv/m2f
         u17eeixYBGW7stk7Q2DjijJdD3Ok2vM4FlTxAwoMKfvcdN61JbJMeUUZJjFEx7ebopRC
         6p2QZyO5M6kidGpf26d9BnG/VX0reitmagz95ckCTCRAdKCIGRxjOqUtqUDu217AJ5BM
         F+Tg==
X-Gm-Message-State: ABuFfogIN/2zDJfRxw4CZDpXDOTZ+U0j27zTFI7MG4YGFV5weScp0arO
        qNa3FIpz2PCMHlGdl7+WCHOW0GMZoxL+otGah5HNVHOxn2bhqDEXIKp2QrQTLmAij4juskHojRg
        qOOWqO3zZjdMN8hJfurE0SrhB75g5M4a2C5LX3L5vQ6vJB/xEfV9zzrj3RA==
X-Google-Smtp-Source: 
 ACcGV62+wp02/h4qAjp9N5OiO3ex2fVz283hZ6GxnpjR1HPVhq7gFPaBIGcklU4EkAJ8SzURZIoQ20aXNm0=
X-Received: by 2002:a24:e48b:: with SMTP id
 o133-v6mr5406027ith.0.1539283617379;
 Thu, 11 Oct 2018 11:46:57 -0700 (PDT)
Date: Thu, 11 Oct 2018 11:46:46 -0700
Message-Id: <20181011184646.154065-1-pshier@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.19.0.605.g01d371f741-goog
Subject: [PATCH] KVM: nVMX: unconditionally cancel preemption timer in
 free_nested
From: Peter Shier <pshier@google.com>
To: kvm@vger.kernel.org
Cc: Peter Shier <pshier@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

There are multiple code paths where an hrtimer may have been started to
emulate an L1 VMX preemption timer that can result in a call to free_nested
without an intervening L2 exit where the hrtimer is normally
cancelled. Unconditionally cancel in free_nested to cover all cases.

Signed-off-by: Peter Shier <pshier@google.com>
Reviewed-by: Jim Mattson <jmattson@google.com>
Reported-by: Jim Mattson <jmattson@google.com>
---
 arch/x86/kvm/vmx.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/arch/x86/kvm/vmx.c b/arch/x86/kvm/vmx.c
index 612fd17be6351..98b8d2d7e7d5a 100644
--- a/arch/x86/kvm/vmx.c
+++ b/arch/x86/kvm/vmx.c
@@ -8450,6 +8450,7 @@ static void free_nested(struct vcpu_vmx *vmx)
 	if (!vmx->nested.vmxon && !vmx->nested.smm.vmxon)
 		return;
 
+	hrtimer_cancel(&vmx->nested.preemption_timer);
 	vmx->nested.vmxon = false;
 	vmx->nested.smm.vmxon = false;
 	free_vpid(vmx->nested.vpid02);
